### ISSUE TITLE
Add DICOM SR Support for additional/common annotations

### ISSUE THREAD

**Issue by dannyrb** [OHIF/Viewers#1215]
I think on first pass we should support:

- Length (done)
- Bidirectional (WIP)
- arrow annotation (…is this doable, could be a single point + a descriptive label?)
- Probe (control point)
- Ellipse
- Rectangle

@pieper has kindly provided some guidance on how to approach these mappings in a discussion with @galelis, @JamesAPetts, and @dannyrb 

@galelis, if you transition off this issue, please provide insights/next-steps in a follow-up comment.

**Comment by dclunie** [comment]
There are 2D (image) relative coordinates in DICOM SR, and 3D (frame of reference e.g., patient, volume, slide).

All of Graphic Types specified for those that are co-planar with a referenced image orientation should be supported (3D types that are not in the same plane as the image would require volume "awareness", which may be out of scope for now):

2D: POINT, MULTIPOINT, POLYLINE, CIRCLE, ELLIPSE
3D: POINT, MULTIPOINT, POLYLINE, POLYGON, ELLIPSE

See:

http://dicom.nema.org/medical/dicom/current/output/chtml/part03/sect_C.18.6.html
http://dicom.nema.org/medical/dicom/current/output/chtml/part03/sect_C.18.9.html

In DICOM SR TID 1500 (which is the template to focus on), see examples of use in these templates:

http://dicom.nema.org/medical/dicom/current/output/chtml/part16/chapter_A.html#sect_TID_1410
http://dicom.nema.org/medical/dicom/current/output/chtml/part16/chapter_A.html#sect_TID_1411
http://dicom.nema.org/medical/dicom/current/output/chtml/part16/chapter_A.html#sect_TID_320

Examples at:

http://dicom.nema.org/medical/dicom/current/output/chtml/part17/sect_RRR.5.html

David

PS. Note that "generic" SR annotation rendering (as opposed to authoring) code may theoretically "ignore" what template is defined to be or appears to be used, and just search the tree for any SCOORD or SCOORD3D content item, and then explore the local neighborhood for children that are images, and parents that describe its context (i.e., any associated description, identification, measurements, etc.). That approach is robust and can support all sorts of SR templates and applications from the generic TID 1500 to Mammography CAD, as well as private templates.

**Comment by dclunie** [comment]
Agree "arrows" should be single point.

**Comment by JamesAPetts** [comment]
Thanks the input David, I thought as much for the arrow.

Yeah the inital scope for this issue would be the in plane annotations that cornerstone supports by default.


Out of plane annotations as well as TID 1411 (?, off the top of my head as I'm on a train, the volumetric measurements for ROIs wil come down the line as separate issues).

**Comment by dannyrb** [comment]
I've created a `Structured Reports` label to tie together related issues. I think we now have enough information to forge a path forward. I've tried to outline a path forward; it's very likely I've stumbled over the correct terminology, but hopefully this conveys things well enough.

---

To @dclunie's note, I think we can achieve different levels of interoperability and functionality. The approach he describes here:

> "generic" SR annotation rendering (as opposed to authoring) code may theoretically "ignore" what template is defined to be or appears to be used, and just search the tree for any SCOORD or SCOORD3D content item, and then explore the local neighborhood for children that are images, and parents that describe its context (i.e., any associated description, identification, measurements, etc.).

May be the best we can do when our application is not the author of the report. or without additional assurances/constraints.

---

# DICOM SR Support for primitive measurements

## Goals

- Improve test coverage and features of the DICOM SR logic in `@ohif/core`
- Expand information made available through services
- Demonstrate best practices around consuming enhanced data services in a practical example (DICOM SR Extension)

## Flow of Data (simplified)

- OHIF Extension registers a `sopClassHandler` for Structured Report SOP Class UIDs
  - [This should function similar to the `loadSegmentation` functionality implemented in the feature branch for our DICOM Seg changes](https://github.com/JamesAPetts/Viewers/blob/vtkjsViewSegmentations/extensions/dicom-seg/src/OHIFDicomSegSopClassHandler.js#L53-L65)
    - Don't create thumbnails for DICOM SR
    - Parse the most recent DICOM SR
    - Maintain a list of all other (historical) DICOM SRs
      - Similar to how DICOM Seg creates a list of labelmaps we can load asynchronously and switch between
- Parsed DICOM SRs hydrate the [OHIF `MeasurementService`](https://github.com/OHIF/Viewers/issues/1281)
- CornerstoneTools syncs it's own imageIdSpecific toolstate with MeasurementService information; Hydrated ToolState is for "enabled" (read only) tools

### Context

We currently parse and render the contents of a DICOM SR by doing the following:

```js
import * as dcmjs from 'dcmjs';

const { DicomMetaDictionary, DicomMessage } = dcmjs.data;

  setContentFromByteArray(byteArray) {
    const arrayBuffer = byteArray.buffer;
    const dicomData = DicomMessage.readFile(arrayBuffer);
    const dataset = DicomMetaDictionary.naturalizeDataset(dicomData.dict);
    dataset._meta = DicomMetaDictionary.namifyDataset(dicomData.meta);

    const mainData = getMainData(dataset);
    const contentSequence = getContentSequence(dataset);
    const content = (
      <>
        {mainData}
        {contentSequence}
      </>
    );

    this.setState({
      content,
    });
  }
```

The `OHIFDicomHtmlViewport` uses the `DicomLoaderService` to get the `byteArray` for a given DisplaySet. It then sets that `byteArray` as a prop for the Viewport. The Viewport uses DCMJS, as outlined above, to parse and display the content in a more or less raw format.

While this has worked as a quick demonstration, it's not as flexible or useful as we would like it to be.

## Blockers

- We should complete the decoupling of the "MeasurementsPanel" extension, so we can remove it for this feature showcase. This extension could reuse the components, but should provide them itself, and marry them with DICOM SR functionality in `@ohif/core`
- Initial completion of the MeasurementService
- The MeasurementService itself is blocked by a few minor changes to CornerstoneTools that allow it to tie annotations to MeasurementService entries

## Phases

I'm afraid if we try to tackle this all in one go that it will stagnate in limbo. Here are some reasonable milestones we can complete to add value and move the bar forward:

### Shift where/how we query and store DICOM SR information

The MeasurementService works well for our critical path, but it doesn't encapsulate all of the information contained in an SR and it's not immutable. We likely want to extend the existing `studyMetadataManager` that contains instances of `OHIFStudyMetadata`. This is our immutable representation, and can be used to hydrate the `MeasurementService`

1. Add a `sopClassHandler` for DICOM SR
2. No longer show SR thumbnails
3. Register/add a copy of the SRs information to a "data service" like the `studyMetadataManager`
4. Replace the `@ohif/extension-dicom-html` logic with a query to the data service that fetches and displays content similar to what is already shown

### Display DICOM SR measurements in a table and on a 2D canvas

The simplest approach here is likely what @dclunie outlined above. We may want to limit the (initial) supported measurements to SCOORD as they'll have a reference image and are easier to map to Cornerstone's `imageIdSpecificToolState`. While we can support SCOORD3D, how that impacts rendering and tabular data is a UX concern that requires more thought.

1. Display a list of SRs in a DICOM SR side panel
    - Order from most recent to oldest
2. Similar to DICOM Seg functionality, hydrate the measurements service with the most recent report. Selecting a previous report should swap out Measurements specific to one report with the newly selected reports and trigger a re-render. Some meta summary for each report should exist.
3. It is problematic/hazardous to assume, without further guarantees, that the parsed measurements can be safely mapped to "edit" tools. Until/unless we have more information, we should use one of the following strategies for render-only behavior:
    - A single tool that, given some schema, can render `POINT`, `MULTIPOINT`, `POLYLINE`, `CIRCLE`, and `ELLIPSE` value types for a given imageId
        - If we're feeling ambitious, this could be the beginning of a tool state SVG renderer that allows interop between VTK and Cornerstone
    - OR we duplicate existing tools (length, freehand, circle, ellipse) with alternative tool names in the "enabled" (read-only) mode that we designate as renderers for mapped measurements

This may sound more straightforward than it is. It's worth looking over the various templates and how they group measurements. We'll need to account for related measurements that are linked by UID (and span measurement groups), as well as a few other scenarios. This may impact how we display information in our table. Some measurements will explicitly map to a POINT or POLYLINE, while others are implied or derived from neighboring information.

### When a TID 1500 Report originated from a supported OHIF version/schema, map to edit tools

We currently accomplish this by keying off of a Measurement's "Code Meaning". For example, we encode a POLYLINE that has a CodeMeaning of `Length`, then assume if we filter/match on that string key, the measurement works with the cornerstone-tools Length Tool. (This currently takes place in a [DCMJS "adapter"](https://github.com/dcmjs-org/dcmjs/blob/7d89d1fcd76953eed27a8eaa1fb75ab449f87c19/src/adapters/Cornerstone/MeasurementReport.js#L173), not within this monorepo).

The [TID 1500](http://dicom.nema.org/medical/dicom/current/output/chtml/part21/sect_A.7.2.html) includes [Author Observer Sequence](https://dicom.innolitics.com/ciods/enhanced-sr/sr-document-general/0040a078). This would allow us to encode information about the application that generated the report and to make further assumptions about the encoded data. For example, if we see that there is an Author Observer (there may be a better field for this) that has:

- (0018,1020) Software Version(s) X
- (0008,0080) Institution Name Y
- etc.

We may be able to reasonably assert that because our application supports Application Y and Software Version X that the encoded measurements are no longer generic, but rather, map `1:1` to our application's tools.

The biggest change here is identifying what information we key off of, and how we might go about supporting schemas as we continue to version/grow what we do support. Ideally, this would be extensible as we grow this to support multi-step and composite tools, but I'm likely getting ahead of myself.

1. Spell out the information we set when persisting information and restoring it
2. Create new fields in the Data Service that help indicate the OHIF version/schema associated, if any, with the SR
3. If the application supports the defined version/schema, map tool state to edit tools instead of the generic render tools

This may sound more straightforward than it is. There are several "gotchas" that may not be immediately obvious. For example:

There is no "rectangle" value type. A rectangle ROI would be encoded as a POLYLINE. Rendering a POLYLINE is simple enough, but if we need to know this should map to a "rectangle tool", we would need to look at the points more closely to identify that. Knowing when a square POLYLINE maps to a square tool, a rectangle tool, or a freehand roi tool is an interesting problem.

POLYLINEs are only "closed" if there first point is the same as their end point.

POLYLINEs do not guarantee a "winding" order. You cannot assume that the points are read from left to right, or right to left. This makes encoding an "arrow" with an end point an interesting problem. We may need to use a single point (end point), and a magnitude/direction

### Round trip persistence

Today, we only support this with the Length tool, and in a way that may not be interoperable with other programs that have support for DICOM SR TID 1500s.

### Support for persisting tools outside the scope of primitive value types (far future)

We will likely need to create a custom "Code Scheme Designator", "Code Meaning"s, and "Code Value"s when persisting tool state for tools that can't be encoded using a single value type. A GitHub repository (and npm package) can describe these concepts, and create an avenue for other systems that would like to consume the data beyond simply rendering/listing the values.

I am currently unsure of how we might provide a path to add these dynamically or ad-hoc through extensions. Interoperability beyond our application may only be possible with published definitions and a concerted effort.

## Not Discussed

#### SCOORD3D Support

While we have a frame of reference tool state manager in cornerstone tools, it's not commonly used. For most measurements (created in OHIF), we'll want to tie them to a specific instance. As we broaden VTK annotation support, SCOORD3D will be more in demand

#### Pathway for rendering w/ VTK Viewports

...

#### Automatic mapping

While we _could_ make optimistic guesses to map SCOORD and SCOORD3D to edit tools, the outcome is likely too hazardous to pursue. A future enhancement could be made to allow manually transferring the state to a tool that supports that same general schema, but with a very blatant warning.

If there is an obvious pattern where MANUFACTURER maps to _these tools_, we could build this in as a more robust feature where users can setup the automatic mappings.

Akin to using ImagerPixelspacing or calibrating an image's pixel spacing. "You can do this, but at your own risk."
  
I am doubtful we will have the bandwidth to tackle this anytime soon, but I believe it's worth listing as it may catch the right financier's eye <3

## Parting Words

Please feel free to push on any/all of this. This is my best guess at our path forward with a few sparse implementation details peppered in. I'm fairly new to DICOM SR, and curious to hear how other implementers are tackling these same problems

**Comment by JamesAPetts** [comment]
> No longer show SR thumbnails

> Some meta summary for each report should exist.

Just so we are on the same page, the HTML render will still be available, just accessed from elsewhere?

> A single tool that, given some schema, can render POINT, MULTIPOINT, POLYLINE, CIRCLE, and ELLIPSE value types for a given imageId

> OR we duplicate existing tools (length, freehand, circle, ellipse) with alternative tool names in the “enabled” (read-only) mode that we designate as renderers for mapped measurements

Duplicating the tools is overkill I reckon? We could add a locked state for all toolState that makes them immutable. This may be too intrusive to CST (unless this a feature others would value?). I think the “SR measurement render only tool” you suggested is better.

>If we’re feeling ambitious, this could be the beginning of a tool state SVG renderer that allows interop between VTK and Cornerstone

I feel like this is a mix of concerns that will only confuse intergation, and is better left as a separate ticket/project later.

> Would we ever want to show multiple SRs measurements at the same time?

Ideally yes, if they contain different data and not just a “corrected” version, but how could we possibly work this out? :thinking_face: . I don’t think we can. Maybe we have to limit to only one SR per series (even that is a little complicated as SRs can reference lots of images from a study I believe). Of course other than a PWA, a more complicated backend could triage time versioned SR.

> We will likely need to create a custom “Code Scheme Designator” …. I am currently unsure of how we might provide a path to add these dynamically or ad-hoc through extensions.

I think the OHIF standard we make must be consumed and added to a Code Scheme library. An extension could add its on CSD + definitions to this library (e.g. an extension adds a tool and how to store it?).

**Comment by dannyrb** [comment]
> Just so we are on the same page, the HTML render will still be available, just accessed from elsewhere?

The HTML Viewport Extension would use the `studyMetadataManager` or similar service to query for the data it needs to render similar output. Setting it as a viewport, if we remove SR thumbnails, would require an action to be dispatched to Redux.

Improved programmtic control of the Viewport Grid and Viewports is something we have slated.

> Duplicating the tools is overkill I reckon?

```js
csTools.addTool(csTools.LengthTool, { name: 'DicomSrLength' });
csTools.setToolEnabled('DicomSrLength');
```

But I agree. I think a render only tool would be cleaner.

> I feel like this is a mix of concerns that will only confuse integration, and is better left as a separate ticket/project later.

I agree. A pipe dream of mine.

> I think the OHIF standard we make must be consumed and added to a Code Scheme library.

👍

**Comment by galelis** [comment]
Regarding the way we display the SRs and how to get old SRs

>Just so we are on the same page, the HTML render will still be available, just accessed from elsewhere?

As SR is also used for more things other than just measurements, we could filter the DICOM SR with measurements from thumbnail, maybe using  `(0018,1020) Software Version(s) X ` or `(0008,0080) Institution Name` or other tag. And display the non-measurement ones as we use to do now.

> Would we ever want to show multiple SRs measurements at the same time?

We could have a runtime load of a new SR, similar to what we have on `series-list` <-> `viewports`, maybe drag and drop into the table or a simple drop down with old SRs (this was suggested by Steve when I first add measurement table on Meteor version) 

> Maybe we have to limit to only one SR per series (even that is a little complicated as SRs can reference lots of images from a study I believe)

Now a days, we use dcmjs and store one SR per Study, doing one by series, I don't personally don't see any gain on having it by series. 

Regarding the way we display the tools, I like the idea of render only tool. 

Have you guys thought about tool versioning and synchronization with what we have on a DICOM SR? I mean, what if the data we stored in the past, new app does not know how to render anymore, like the structure has changed..

**Comment by fedorov** [comment]
Wow, that's a lot of stuff guys! I have to say I could not follow all of the points in the discussion due to lacking understanding of the OHIF internals perhaps. My overall suggestion would be to consider focusing on specific use cases, find a very specific customer that will be invested in using this functionality and provide feedback to guide development. Just one TID 1500 template has so many capabilities and usage flavors that even exhaustively covering those is a huge project! 

On a positive side, I guess (considering the "I am doubtful we will have the bandwidth to tackle this anytime soon" comment!), planar annotations are not an immediate priority for the IDC project. Our current target is development of the MVP (see our plans in the IDC MVP document linked from [here](https://imagingdatacommons.github.io/info/)), which will provide static view of a subset of TCIA collections. None of those collections (with perhaps the exception of the annotations produced by CCC for some of them) have planar annotations. It is much more important for IDC to be able to handle measurements and SR objects derived from volumetric segmentations - plenty of datasets of that kind, and more on the way.

I promised to @pieper and @swederik that I will submit all issues that I encountered and also create issues for functionality we need for IDC MVP. We can discuss those after that in a call to hopefully bring everyone to the same page.

**Comment by dannyrb** [comment]
Thank you for your thoughts, @fedorov ^_^

> It is much more important for IDC to be able to handle measurements and SR objects derived from volumetric segmentations - plenty of datasets of that kind, and more on the way.

**Comment by pieper** [comment]
Also just for context: the effort to serialize and deserialize OHIF annotations is not just for IDC, although there's great synergy with the efforts to use SR for this purpose especially given the guidance from @dclunie 's white paper.

Currently we have IDC:candidate and IDC:priority labels, but maybe this would fall into something like an IDC:collaboration category (don't want to invent too many labels, but if needed we can).

**Comment by fedorov** [comment]
@pieper I agree, and adding "IDC:collaboration" makes sense to me.

**Comment by nutzlastfan** [comment]
> Akin to using ImagerPixelspacing or calibrating an image's pixel spacing. "You can do this, but at your own risk."

I am doubtful we will have the bandwidth to tackle this anytime soon, but I believe it's worth listing **as it may catch the right financier's eye** <3

Hi, we would be very much interested in the possibility to calibrate measurement in an image.
(some pictures are taken with objects like metal balls or rulers which are used as reference for all later measurements on the picture) So if financing is an issue maybe we can work something out.
If this is of interest please contact me directly.

> > ... "generic" SR annotation rendering (as opposed to authoring) code may theoretically "ignore" what template is defined to be or appears to be used, and just search the tree for any SCOORD or SCOORD3D content item, and then explore the local neighborhood for children that are images, and parents that describe its context (i.e., any associated description, identification, measurements, etc.). - @dclunie
> 
> This is a good note. When our application is not the author of the report, and without additional assurances/constraints, this may be the best we can do without introducing hazards.
> 
> ## Overview
> ### Current approach
> We parse and render the contents of a DICOM SR by doing the following:
> 
> ```js
> import * as dcmjs from 'dcmjs';
> 
> const { DicomMetaDictionary, DicomMessage } = dcmjs.data;
> 
>   setContentFromByteArray(byteArray) {
>     const arrayBuffer = byteArray.buffer;
>     const dicomData = DicomMessage.readFile(arrayBuffer);
>     const dataset = DicomMetaDictionary.naturalizeDataset(dicomData.dict);
>     dataset._meta = DicomMetaDictionary.namifyDataset(dicomData.meta);
> 
>     const mainData = getMainData(dataset);
>     const contentSequence = getContentSequence(dataset);
>     const content = (
>       <>
>         {mainData}
>         {contentSequence}
>       </>
>     );
> 
>     this.setState({
>       content,
>     });
>   }
> ```
> 
> A `sopClassHandler` is registered that creates displaySets for DICOM SR SOP Class UIDs. The `OHIFDicomHtmlViewport` uses the `DicomLoaderService` to get the `byteArray` for a given DisplaySet. It then sets that `byteArray` as a prop for the Viewport. The Viewport uses DCMJS, as outlined above, to parse and display the content in a more or less raw format. When mapping a measurement to tool state, we use the parsed dataset with a DCMJS adapter.
> 
> ### Issues with current approach
> * Extensions are forced to have a hard dependency on `DCMJS` and the `DicomLoaderService`
> * Extensions are required to have an intimate knowledge of DICOM SR to extract/persist data
> * Extensions that need the same information would dispatch multiple (duplicate) network requests
> * Mapping SR measurements to tools can be done in a more interoperable way
> * Mapping SR measurements to CSTools' State bypasses the platform and makes CSTools the source of truth
> * We are unable to display measurements created by other sources
> 
> ### Goals
> * Centralize DICOM SR Parsing / loading
> * Broad support for rendering DICOM SR measurements
> * Ability to _create_ interoperable DICOM SRs that have edit support for primitives
> * Ease burden of DICOM SR knowledge for extension authors
> * Ease ability to create new extensions that can leverage information in DICOM SRs
> * (Showcase) Best practices around consuming enhanced data services in a practical example (DICOM SR Extension)
> 
> ### Blockers
> * MVP completion of the MeasurementService
> * We should complete the decoupling of the "MeasurementsPanel" extension, so we can remove it for this feature showcase. This extension could reuse the components, but should provide them itself, and marry them with DICOM SR functionality in `@ohif/core`
> * Current `studyMetadataManager` usage is convoluted
> 
> ## Phases
> This task is too large to tackle in one go. Here are some reasonable milestones we can use to iteratively add value:
> 
> ### Shift where/how we query and store DICOM SR information
> The MeasurementService works well for our critical path, but it doesn't encapsulate all of the information contained in an SR and it's not immutable. We likely want to extend the existing `studyMetadataManager` that contains instances of `OHIFStudyMetadata`. It can be used to hydrate the `MeasurementService`.
> 
> 1. Add a `sopClassHandler` for DICOM SR
>    
>    * No longer show SR thumbnails
> 2. Register/add a copy of the SRs information to a "data service" like the `studyMetadataManager`
> 3. Replace the `@ohif/extension-dicom-html` logic with a query to the data service that fetches and displays content similar to what is already shown
> 
> ### Display DICOM SR measurements in a table and on a 2D canvas
> The simplest approach here is likely what @dclunie outlined above. We may want to limit the (initial) supported measurements to SCOORD as they'll have a reference image and are easier to map to Cornerstone's `imageIdSpecificToolState`. While we can support SCOORD3D, how that impacts rendering and tabular data is a UX concern that requires more thought.
> 
> 1. Display a list of SRs in a DICOM SR side panel
>    
>    * Order from most recent to oldest
> 2. Similar to DICOM Seg functionality, hydrate the measurements service with the most recent report. Selecting a previous report should swap out Measurements specific to one report with the newly selected reports and trigger a re-render. Some meta summary for each report should exist.
> 3. It is problematic/hazardous to assume, without further guarantees, that the parsed measurements can be safely mapped to "edit" tools. Until we have more information, we should use one of the following strategies for render-only behavior:
>    
>    * A single tool that, given some schema, can render `POINT`, `MULTIPOINT`, `POLYLINE`, `CIRCLE`, and `ELLIPSE` value types for a given imageId
>      
>      * If we're feeling ambitious, this could be the beginning of a tool state SVG renderer that allows interop between VTK and Cornerstone
>    * OR we duplicate existing tools (length, freehand, circle, ellipse) with alternative tool names in the "enabled" (read-only) mode that we designate as renderers for mapped measurements
> 
> This may sound more straightforward than it is. It's worth looking over the various templates and how they group measurements. We'll need to account for related measurements that are linked by UID (and span measurement groups), as well as a few other scenarios. This may impact how we display information in our table. Some measurements will explicitly map to a POINT or POLYLINE, while others are implied or derived from neighboring information.
> 
> #### Notes:
> * Do timepoints play a part here?
> * Would we ever want to show multiple SRs measurements at the same time?
> 
> ### When a TID 1500 Report originated from a supported OHIF version/schema, map to edit tools
> We currently accomplish this by keying off of a Measurement's "Code Meaning". For example, we encode a POLYLINE that has a CodeMeaning of `Length`, then assume if we filter/match on that string key, the measurement works with the cornerstone-tools Length Tool. (This currently takes place in a [DCMJS "adapter"](https://github.com/dcmjs-org/dcmjs/blob/7d89d1fcd76953eed27a8eaa1fb75ab449f87c19/src/adapters/Cornerstone/MeasurementReport.js#L173), not within this monorepo).
> 
> The [TID 1500](http://dicom.nema.org/medical/dicom/current/output/chtml/part21/sect_A.7.2.html) includes [Author Observer Sequence](https://dicom.innolitics.com/ciods/enhanced-sr/sr-document-general/0040a078). This would allow us to encode information about the application that generated the report and to make further assumptions about the encoded data. For example, if we see that there is an Author Observer (there may be a better field for this) that has:
> 
> * (0018,1020) Software Version(s) X
> * (0008,0080) Institution Name Y
> * etc.
> 
> We may be able to reasonably assert that because our application supports Application Y and Software Version X that the encoded measurements are no longer generic, but rather, map `1:1` to our application's tools.
> 
> The biggest change here is identifying what information we key off of, and how we might go about supporting schemas as we continue to version/grow what we do support. Ideally, this would be extensible as we grow this to support multi-step and composite tools, but I'm likely getting ahead of myself.
> 
> 1. Spell out the information we set when persisting information and restoring it
> 2. Create new fields in the Data Service that help indicate the OHIF version/schema associated, if any, with the SR
> 3. If the application supports the defined version/schema, map tool state to edit tools instead of the generic render tools
> 4. Support round trip persistence
> 
> This may sound more straightforward than it is. There are several "gotchas" that may not be immediately obvious. For example:
> 
> There is no "rectangle" value type. A rectangle ROI would be encoded as a POLYLINE. Rendering a POLYLINE is simple enough, but if we need to know this should map to a "rectangle tool", we would need to look at the points more closely to identify that. Knowing when a square POLYLINE maps to a square tool, a rectangle tool, or a freehand roi tool is an interesting problem.
> 
> POLYLINEs are only "closed" if the first point is the same as the end point.
> 
> POLYLINEs do not guarantee a "winding" order. You cannot assume that the points are read from left to right, or right to left. This makes encoding an "arrow" with an end point an interesting problem. We may need to use a single point (end point), and a magnitude/direction
> 
> ### Support for persisting tools outside the scope of primitive value types (far future)
> We will likely need to create a custom "Code Scheme Designator", "Code Meaning"s, and "Code Value"s when persisting tool state for tools that can't be encoded using a single value type. A GitHub repository (and npm package) can describe these concepts, and create an avenue for other systems that would like to consume the data beyond simply rendering/listing the values.
> 
> I am currently unsure of how we might provide a path to add these dynamically or ad-hoc through extensions. Interoperability beyond our application may only be possible with published definitions and a concerted effort.
> 
> ## Not Discussed
> #### SCOORD3D Support
> While we have a frame of reference tool state manager in cornerstone tools, it's not commonly used. For most measurements (created in OHIF), we'll want to tie them to a specific instance. As we broaden VTK annotation support, SCOORD3D support will becomes more pressing.
> 
> #### Pathway for rendering w/ VTK Viewports
> ...
> 
> #### Automatic mapping
> While we _could_ make optimistic guesses to map SCOORD and SCOORD3D to edit tools, the outcome is likely too hazardous to pursue. A future enhancement could be made to allow manually transferring the state to a tool that supports that same general schema, but with a very blatant warning.
> 
> If there is an obvious pattern where MANUFACTURER maps to _these tools_, we could build this in as a more robust feature where users can setup the automatic mappings.
> 
> Akin to using ImagerPixelspacing or calibrating an image's pixel spacing. "You can do this, but at your own risk."
> 
> I am doubtful we will have the bandwidth to tackle this anytime soon, but I believe it's worth listing as it may catch the right financier's eye <3
> 
> ## Parting Words
> Please feel free to push on any/all of this. This is my best guess at our path forward with a few sparse implementation details peppered in. I'm fairly new to DICOM SR, and curious to hear how other implementers are tackling these same problems.
> 
> Note: I've created a [`Structured Reports :memo:`](https://github.com/OHIF/Viewers/labels/Structured%20Reports%20%3Amemo%3A) label to tie together related issues.

> Akin to using ImagerPixelspacing or calibrating an image's pixel spacing. "You can do this, but at your own risk."

I am doubtful we will have the bandwidth to tackle this anytime soon, but I believe it's worth listing **as it may catch the right financier's eye** <3

Hi, we would be very much interested in the possibility to calibrate measurement in an image.
(some pictures are taken with objects like metal balls or rulers which are used as reference for all later measurements on the picture) So if financing is an issue maybe we can work something out.
If this is of interest please contact me directly.

**Comment by dannyrb** [comment]
:wave: Hi @nutzlastfan, as far as I can tell, you don't have public contact information associated with your GitHub account. Radical Imaging staff comprises the bulk of our core team. I recommend reaching out to:

Rob Lewis (Owner)
rob@radicalimaging.com

Erik Ziegler (Director of R&D)
erik@radicalimaging.com

**Comment by igoroctaviano** [comment]
> I think on first pass we should support:
> 
> * Length (done)
> * Bidirectional (WIP)
> * arrow annotation (…is this doable, could be a single point + a descriptive label?)
> * Probe (control point)
> * Ellipse
> * Rectangle
> 
> @pieper has kindly provided some guidance on how to approach these mappings in a discussion with @galelis, @JamesAPetts, and @dannyrb
> 
> @galelis, if you transition off this issue, please provide insights/next-steps in a follow-up comment.

@JamesAPetts this ticket was recently moved to IDC priority. 
Looks like some progress was done here #1811. 
- Do we need to update this issue considering those changes?
- There are things missing / tech debt that can be tackled here?
- Considering the tools we have in the viewer, the missing ones are Angle and Rectangle?

**Comment by dclunie** [comment]
Glad to see there may be progress on this.

Since I last commented, I drafted this paper on DICOM SR for communicating planar annotations, though I haven't looked at it in a while: [http://docs.google.com/document/d/1bR6m7foTCzofoZKeIRN5YreBrkjgMcBfNA7r9wXEGR4/](http://docs.google.com/document/d/1bR6m7foTCzofoZKeIRN5YreBrkjgMcBfNA7r9wXEGR4/)

**Comment by robertsoakes** [comment]
@dannyrb @JamesAPetts Is there any way to see what is actively being worked on for this issue? My company is actively working on some extensions for the viewer to be used in an orthopedic patient matched process, and need to implement support for some of the additional tools.

We have a developer (@MaksFasolka) who is currently researching what needs to be done, but we're not clear on the next steps. Is there a member of the Radical Imaging team who is currently responsible for creating the MeasurementService or implementing support for additional tools?

**Comment by Punzo** [comment]
This ticket is currently assigned to @igoroctaviano and it is an IDC collaboration/priority (@pieper @fedorov)

**Comment by fedorov** [comment]
@igoroctaviano I put samples of additional planar annotations as follows:
* gs://tcia-idc-datareviewcoordination/issue-27-prostatex contains a study from ProstateX collection, with an SR that has a single-point SCROORD3D annotation
* gs://tcia-idc-datareviewcoordination/issue-28-petctlungdx has a study from PET-CT-LungDx collection, with several SCOORD bounding box annotations

Can you use those in your tests?

I also imported those into the `projects/idc-tcia/locations/us-central1/datasets/tcia-idc-datareviewcoordination/dicomStores/planar_annotations` DICOM store, but somehow I do not see them in the OHIF sandbox - they files are in the bucket, and there were no import errors, but I don't know if it's an OHIF issue, or a Google Healthcare issue.

**Comment by igoroctaviano** [comment]
> @igoroctaviano I put samples of additional planar annotations as follows:
> 
> * gs://tcia-idc-datareviewcoordination/issue-27-prostatex contains a study from ProstateX collection, with an SR that has a single-point SCROORD3D annotation
> * gs://tcia-idc-datareviewcoordination/issue-28-petctlungdx has a study from PET-CT-LungDx collection, with several SCOORD bounding box annotations
> 
> Can you use those in your tests?
> 
> I also imported those into the `projects/idc-tcia/locations/us-central1/datasets/tcia-idc-datareviewcoordination/dicomStores/planar_annotations` DICOM store, but somehow I do not see them in the OHIF sandbox - they files are in the bucket, and there were no import errors, but I don't know if it's an OHIF issue, or a Google Healthcare issue.

Great, thanks. Regarding the bounding box, It's probably because the last related PR only added support for length and bidirectional tools.

**Comment by fedorov** [comment]
As discussed with @Punzo today, let's wait until @igoroctaviano is back on IDC in December to continue this work.

**Comment by fedorov** [comment]
As discussed with Markus, we should consider if the code used in Slim for mapping SCOORD3D to dcmjs. Davide to discuss with Markus.

**Comment by fedorov** [comment]
@igoroctaviano we were trying to use this feature, but it is not clear what is being used to define association between the SR with the point annotation and the series it is possible to overlay this point. Can you please help?

First, I would expect that it should be possible to overlay the point over any series that shares `FrameOfReferenceUID`, but it appears that it would show up only for a specific series.

Second, selection of the series over which it is possible to render the point is unclear. In our case, we have multiple MR series within the study, and when the point under "Measurements" tab is clicked, a specific series is loaded with the point (arrow) overlay. It is not the series that we want it to be overlayed on. We then tried setting `ReferencedStudySequence > ReferencedSOPInstanceUID`, in addition to `PertinentOtherEvidenceSequence  > ReferencedSeriesSequence > SeriesInstanceUID` (same pattern as in the sample that does work here: https://idc-sandbox-000.firebaseapp.com/projects/idc-sandbox-000/locations/us/datasets/idc-dicom-test-inventory/dicomStores/test-samples/study/1.3.6.1.4.1.14519.5.2.1.7310.5101.860473186348887719777907797922) where that referenced SOPInstanceUID is from a series we want to overlay the point, but this does not help - same series is being shown. Next we tried removing all of the MR series from the study except for the one we want to see the overlay, but the result was that the "Measurements" tab disappeared altogether. 

cc: @PaWeRe

**Comment by yoni-bme** [comment]
Can I get the file

**Comment by igoroctaviano** [comment]
> @igoroctaviano we were trying to use this feature, but it is not clear what is being used to define association between the SR with the point annotation and the series it is possible to overlay this point. Can you please help?
> 
> First, I would expect that it should be possible to overlay the point over any series that shares `FrameOfReferenceUID`, but it appears that it would show up only for a specific series.
> 
> Second, selection of the series over which it is possible to render the point is unclear. In our case, we have multiple MR series within the study, and when the point under "Measurements" tab is clicked, a specific series is loaded with the point (arrow) overlay. It is not the series that we want it to be overlayed on. We then tried setting `ReferencedStudySequence > ReferencedSOPInstanceUID`, in addition to `PertinentOtherEvidenceSequence > ReferencedSeriesSequence > SeriesInstanceUID` (same pattern as in the sample that does work here: https://idc-sandbox-000.firebaseapp.com/projects/idc-sandbox-000/locations/us/datasets/idc-dicom-test-inventory/dicomStores/test-samples/study/1.3.6.1.4.1.14519.5.2.1.7310.5101.860473186348887719777907797922) where that referenced SOPInstanceUID is from a series we want to overlay the point, but this does not help - same series is being shown. Next we tried removing all of the MR series from the study except for the one we want to see the overlay, but the result was that the "Measurements" tab disappeared altogether.
> 
> cc: @PaWeRe

@rodrigobasilio2022 will test the dataset in v3 first and then estimate effort for v2.

**Comment by rodrigobasilio2022** [comment]
@fedorov, What do you expect to see in case of the MR prostate? I am testing a solution and this is the result. Its not finished as it doesnt get the projection of the point in the image slice, but this is what I getright now.


![image](https://github.com/OHIF/Viewers/assets/114958722/c50ec2b3-e899-4753-9b61-7d0b7b150bb2)

**Comment by rodrigobasilio2022** [comment]
CT case
![image](https://github.com/OHIF/Viewers/assets/114958722/cb9ab184-9f9d-4d27-bf9c-e357eefc0328)

**Comment by fedorov** [comment]
Looks like a great start! What is the dashed line supposed to do? Should there be text annotation at the end of the dashed line?

**Comment by fedorov** [comment]
> Its the default behaviour. I will change that.

Oh, no need to change it yet - we should just discuss what text to show, but I would want to review the samples first. Are you testing this against the O10 and O11 tests in IDC Test inventory?

> One way to facilitate this task, if I understanding correctly, is to modify the python code tha generate this SRs so I can write more compatible SRs. Do you agree with this direction?

To the best of my knowledge, the samples we have are compliant with the DICOM standard, and that should be the only thing that should matter. If the way dcmjs writes those SRs is different, we should review dcmjs code and confirm it generates proper SR objects.

**Comment by rodrigobasilio2022** [comment]
I am using this database from Dropbox https://www.dropbox.com/s/98rylgt25b2sm9r/planar_annotations.zip?dl=0

**Comment by fedorov** [comment]
Let's switch to using the test I mentioned above from the testing inventory.

**Comment by fedorov** [comment]
@rodrigobasilio2022 what is the current status of this issue?

**Comment by rodrigobasilio2022** [comment]
Working today in this issue. I will release a PR today with that

**Comment by dclunie** [comment]
@rodrigobasilio2022, FYI, even though they aren't in the TCIA DICOM database and IDC hasn't ingested them, the [Crowds Cure Cancer](https://www.crowds-cure.org/) DICOM SR files that consistent of unidimensional (RECIST-style long axis length) and bi-dimensional (WHO-style long and short axis lengths) measurements are available for testing and can be downloaded manually from TCIA for the [2017](https://doi.org/10.7937/K9/TCIA.2018.OW73VLO2) and [2018](https://doi.org/10.7937/TCIA.2019.yk0gm1eb) years.

You could select some representative samples with images from the public access collections and add them to the inventory of tests.

**Comment by rodrigobasilio2022** [comment]
PR opened. I will fix some bugs found in the tests and test some Crowds Cure Cancer images. After that assign the PR to the reviewers

**Comment by sedghi** [comment]
Done in master

**Comment by fedorov** [comment]
We found quite a few issues with the current implementation based on the review at the last IDC-Radical coordination meeting. Even if those have been fixed, I have not heard an update from @rodrigobasilio2022 or @igoroctaviano about it. So we should either reopen this issue, or create a new one. Until a new issue is created, I am reopening this one.

**Comment by sedghi** [comment]
Can we create a new one that is relevant and only include the remaining bugs?

**Comment by rodrigobasilio2022** [comment]
Its not a bug itself in the implementation, but in how the data is presented in Measurement Panel. @fedorov , correct me if i am wrong, but the want to filter the SRs in the panel to the ones related to the active viewport, right?

**Comment by igoroctaviano** [comment]
> Its not a bug itself in the implementation, but in how the data is presented in Measurement Panel. @fedorov , correct me if i am wrong, but the want to filter the SRs in the panel to the ones related to the active viewport, right?

@rodrigobasilio2022 

**Scenarios:**
**Given** I'm a user
**When** I open a study with SCOORD3D points
**Then** The measurement panel displays the exact number of measurements present in the SR 

**Given** I'm a user
**When** I select a series in the gallery (activating it) 
**And** I select a measurement in the right panel (measurement panel)
**Then** The measurement should be rendered on the active series, jumping to the slice where the measurement was rendered (If the active series is not the referenced series in the SR, the measurement should be displayed on that series if it's close enough given a tolerance value)

So the idea here is to parse the SR, display the measurements in the measurement panel (the exact number of measurements present in the SR, no duplication allowed) and when the user clicks on a given measurement, that measurement is displayed on the **active** viewport (the viewer jumps to that slice). This means that the measurement could be rendered on any series if it shares a frame of reference and is close enough to a particular slice given a tolerance value.

**Comment by fedorov** [comment]
Thank you @igoroctaviano - this is exactly correct.

**Comment by rodrigobasilio2022** [comment]
This scenario has a problem, if there is no slice near to the SCOORD3D points in the active viewport, nothing will show. What if we change this scenario a little bit, by filtering the Measurements Table based on the current active viewport. If there is annotations they will be displayed, if not the measurements panel will be empty.

**Comment by igoroctaviano** [comment]
> This scenario has a problem, if there is no slice near to the SCOORD3D points in the active viewport, nothing will show. What if we change this scenario a little bit, by filtering the Measurements Table based on the current active viewport. If there is annotations they will be displayed, if not the measurements panel will be empty.

For now, we can focus on the happy path and later decide on the case you mentioned. 

For the future, If the measurement is not fit for the active viewport it's better to have it disabled (grayed out) than filtered out. This way the user knows that the content of the SR is there but it's not valid because of reason X. Maybe later we could even allow the user to change/tune the tolerance value but this is all future things.

**Comment by fedorov** [comment]
Yes, I agree with @igoroctaviano. Showing a subset of items from SR based on what can be visualized would be very misleading.

> What if we change this scenario a little bit, by filtering the Measurements Table based on the current active viewport. If there is annotations they will be displayed, if not the measurements panel will be empty.

Instead, we could explore a variety of other ways to communicate this to the user (in addition to disabling, as suggested by Igor):
* show warning sign next to the items that do not map to any slice
* show projection to the closest slice, clearly indicating by the appearance of the annotation that it is a projection to a slice beyond the threshold (for example, by displaying warning sign next to the point)

I would implement the easiest of those ideas for now, and revisit later.

**Comment by rodrigobasilio2022** [comment]
![image](https://github.com/OHIF/Viewers/assets/114958722/a0057cb0-4277-448b-972c-271376e8fca6)

Updated the code with this features. Please review. There one more change to do. Its the auto hydration issue, to avoid popping up the infamous dialog box

**Comment by igoroctaviano** [comment]
> ![image](https://user-images.githubusercontent.com/114958722/276279595-a0057cb0-4277-448b-972c-271376e8fca6.png)
> 
> Updated the code with this features. Please review. There one more change to do. Its the auto hydration issue, to avoid popping up the infamous dialog box

Are you updating a different env? I tried with https://d3w0jbfdn2j3pj.cloudfront.net/projects/idc-sandbox-000/locations/us/datasets/idc-dicom-test-inventory/dicomStores/test-samples/study/1.3.6.1.4.1.14519.5.2.1.7310.5101.860473186348887719777907797922

### First series (measurement not active for viewport)
![Screenshot 2023-10-18 at 11 36 09](https://github.com/OHIF/Viewers/assets/13886933/1532cb90-fe1e-4034-90ee-d15dbed84ea7)
### Second series (blank / no measurement although it jumps to a frame)
![Screenshot 2023-10-18 at 11 36 17](https://github.com/OHIF/Viewers/assets/13886933/538ff5c9-f936-4e09-9533-45d5a7a499f7)
### Third series (measurement not active for viewport)
![Screenshot 2023-10-18 at 11 36 31](https://github.com/OHIF/Viewers/assets/13886933/28604944-e979-4a3c-8339-3e845234d84d)

**Comment by rodrigobasilio2022** [comment]
Just series 6, 8, 9 has measurements. For me it goes ok, though I saw clicking with no measurement display, and I click again and it is there. Maybe its a bug in OHIF, but can you try these series and if there is no annotation displaying try again clicking in the same annotation

**Comment by igoroctaviano** [comment]
> Just series 6, 8, 9 has measurements. For me it goes ok, though I saw clicking with no measurement display, and I click again and it is there. Maybe its a bug in OHIF, but can you try these series and if there is no annotation displaying try again clicking in the same annotation

@fedorov is this correct? I thought the measurements should be rendered for other planes too.

**Comment by fedorov** [comment]
> They are translated from SCOORD3D to SCOORD

I do not understand what this means. Translated by whom and when?

**Comment by igoroctaviano** [comment]
> > They are translated from SCOORD3D to SCOORD
> 
> I do not understand what this means. Translated by whom and when?

@fedorov I'll have a meeting with @rodrigobasilio2022 to discuss this. We'll get back to you ASAP.

**Comment by fedorov** [comment]
Thank you Igor. 

@rodrigobasilio2022 the DICOM data received from the server must be shown as received from the server in the DICOM tag viewer. It is absolutely unacceptable to modify and present modified data in the DICOM tag viewer! The details of how you manage internal to the implementation data structures must be completely separate from the DICOM data representation.

**Comment by rodrigobasilio2022** [comment]
Understood. I will change the implementation.

**Comment by fedorov** [comment]
Per discussion today, we will test this once a sandbox instance with these features is deployed.

**Comment by mrceresa** [comment]
Dears, any updates on this?

**Comment by Joeycho** [comment]
@dclunie I hope you remember this issue. Why does Arrow need to be "single" point" rather than "two points"?

Because of this, everytime changing viewport (resizing), arrows are randomly rotated.. Or is there any logics behind it?

**Comment by dclunie** [comment]
Because SR is about semantics, not appearance.

I.e., for an arrow, what matters is what is at the tip, not the direction of the arrow.

**Comment by Joeycho** [comment]
@dclunie Thank you for the prompt response :) Yes, but in my case, a user saves the status with the screenshot, and they load next time. Their expectation is that the arrow should be loaded exactly as they saved. And I think that makes sense as well. And I think if I refer how the line (length) is saved as SR, probably I can edit how arrow is saved and may achieve what I expect. What do you think?

**Comment by sedghi** [comment]
@seandoyle FYI seems like a response to our recent discussion

**Comment by Joeycho** [comment]
@seandoyle @dclunie @sedghi 

I think I found a way to fix this arrow issue and made the following PRs several weeks ago. Fortunately, A point for TID300 can store additional point, which makes the implementation easier. If you can find some issues in the following PRs, it will be great.  

https://github.com/cornerstonejs/cornerstone3D/pull/1450
https://github.com/OHIF/Viewers/pull/4356

**Comment by deepakri201** [comment]
Hi, 

I am using [this viewer](https://viewer.ohif.org/local) to display a structured report with SCOORD3D coordinates. Unfortunately, it does not load in the viewer, and these warnings appear: 

![Image](https://github.com/user-attachments/assets/dbe45dec-0e9b-46ed-bea0-6389146e06a2)

I have attached the SR and the associated MR files. 

Thank you, 

Deepa 

[prostate_point_02_13_25.zip](https://github.com/user-attachments/files/18790628/prostate_point_02_13_25.zip)

**Comment by igoroctaviano** [comment]
> Hi,
> 
> I am using [this viewer](https://viewer.ohif.org/local) to display a structured report with SCOORD3D coordinates. Unfortunately, it does not load in the viewer, and these warnings appear:
> 
> ![Image](https://github.com/user-attachments/assets/dbe45dec-0e9b-46ed-bea0-6389146e06a2)
> 
> I have attached the SR and the associated MR files.
> 
> Thank you,
> 
> Deepa
> 
> [prostate_point_02_13_25.zip](https://github.com/user-attachments/files/18790628/prostate_point_02_13_25.zip)

I think @fedorov could load the scoord3d points using v3 so I'm wondering if something to do with this dataset. @fedorov could you please confirm that?

**Comment by fedorov** [comment]
@deepakri201 can you test with the sample I shared with you earlier as a reference and let us know what you see? Can you also include the bounding box sample and error you observed?

**Comment by deepakri201** [comment]
@fedorov - the sample I shared earlier was actually the one you provided from ProstateX. 

For the bounding box SR, here are the errors I got: 

![Image](https://github.com/user-attachments/assets/1dfdb3e6-331a-442b-bb4c-0475e9591a79)

![Image](https://github.com/user-attachments/assets/232f3135-69f8-423b-9bdb-36f78fda3c03)

The data is at this [dropbox link](https://www.dropbox.com/scl/fi/zxikh4kp53xeg0y7go2i6/OHIF_bboxes_02_13_25.zip?rlkey=koce906ii9js8pg2ewuzc6snu&st=jszu9i08&dl=0).

**Comment by deepakri201** [comment]
@fedorov -- The errors above were actually from the ProstateX sample you shared. 

For the bounding box SR, here are the errors I got: 

![Image](https://github.com/user-attachments/assets/83f07587-3ef7-4bff-958c-36c970dd7edc)

![Image](https://github.com/user-attachments/assets/a940eaf9-b4b1-4718-947f-3a6995591c29)

The data is here at [this dropbox link](https://www.dropbox.com/scl/fi/zxikh4kp53xeg0y7go2i6/OHIF_bboxes_02_13_25.zip?rlkey=koce906ii9js8pg2ewuzc6snu&st=jszu9i08&dl=0).

**Comment by fedorov** [comment]
@igoroctaviano are there any updates to get more clarity on the status of this feature?

**Comment by fedorov** [comment]
@deepakri201 Igor is telling me that the points (and maybe bounding boxes) should now work, but the side panel does not show them up, so that is a bug. If you could test this, it would be great (I will updated the gha v3 instance).

**Comment by deepakri201** [comment]
With the same SR as above, I see that the two measurements appear in some of the scans, but not all. For instance, both measurements appear in Series #6 t2_tse_tra, but only one appears in Series #5 t2_tse_cor, and none in Series #3 t2_tse_sag.

**Comment by igoroctaviano** [comment]
FYI, there’s a distance threshold for projecting these points. @fedorov, should they be rendered on all series? They appear on some, but I’m not sure if they should be on all. Since you’re more familiar with this dataset and may have tested it elsewhere, you might have more insight.

**Comment by fedorov** [comment]
@igoroctaviano my guess is that when we had the panel on the right to click on the point, the viewport would show the image going through the point on click. Without that panel, perhaps it is only possible to see the point if it is within some threshold from the slice, and it is not possible to make the viewport show reformatted slice showing through the point. Is this a possible explanation? 

I think the next step is to fix the regression with the panel showing points.

**Comment by deepakri201** [comment]
Here is a recent bounding box SR with the corresponding CT files. 
(SeriesInstanceUID of CT = 1.2.840.113654.2.55.100267029693057843237393396736558778216)

https://www.dropbox.com/scl/fi/aywi9llbayodv4b8jzksq/ohif_bboxes_05_01_25.zip?rlkey=vdiokr93qc9j7qhs40mvrwebr&st=wcs1wx4w&dl=0

**Comment by fedorov** [comment]
> Here is a recent bounding box SR with the corresponding CT files.
(SeriesInstanceUID of CT = 1.2.840.113654.2.55.100267029693057843237393396736558778216)

To clarify what @deepakri201 said above, the referenced CT is in IDC, and you can fetch the corresponding CT series as follows:

```bash
$ pip install --upgrade idc-index
$ idc download 1.2.840.113654.2.55.100267029693057843237393396736558778216
```

**Comment by wayfarer3130** [comment]
The problem with the SCOORD3D is here: https://dicom.nema.org/medical/dicom/current/output/html/part03.html#sect_A.35.2
We need to use/support SOP Class https://dicom.nema.org/medical/dicom/current/output/html/part03.html#sect_A.35.13
Comprehensive 3d SR

We should also shorten the implementation version name OHIF-VIEWER-2.0.0 to OHIF-3.11.0
Remove CurrentRequestedProcedureEvidenceSequence or perhaps fill it with evidence if relevant (I think remove)
Remove Image Comments
I'm not sure which of those are changes here and which are in Cornerstone3D adapters.

**Comment by fedorov** [comment]
@wayfarer3130 did you mean to add the comment above to this PR instead? https://github.com/dcmjs-org/dcmjs/pull/432

**Comment by wayfarer3130** [comment]
> [@wayfarer3130](https://github.com/wayfarer3130) did you mean to add the comment above to this PR instead? [dcmjs-org/dcmjs#432](https://github.com/dcmjs-org/dcmjs/pull/432)

Yes, moving there.  That PR links to htis one, and I missed the fact that I'd clicked on the reference.

**Comment by wayfarer3130** [comment]
There are a number of issues with the current behaviour for saving/loading.  OHIF will need to continue to load the older object style, but the newer one should be preferred for both storing and parsing/hydrating the measurements.
The SR measurement viewport in OHIF displays the raw measurement positions as polylines, but doesn't load them unless they contain the OHIF specific use of the measurement tracking identifier.   As stated, that should be changed consistently to use the source of measurement, with shared/common CID values to specify the measurement type, and where possible to embed sub-classification specific to OHIF for types that are nearly standard, but which need a small amount of customization to specify exactly how they get handled.   Also, the freeform text should be stored as a free form observation rather than being embedded in finding code.

**Comment by wayfarer3130** [comment]
The recent changes to the measurements panel allows for a great deal more customization than was previously available.  That should allow a separate section to be included only when an SR viewport panel is shown.  That way the list of annotations could be included in the existing panel.  I believe that would address some of the comments about needing to see the list of measurements and being able to navigate among them.

**Comment by fedorov** [comment]
@wayfarer3130 we are creating SRs outside of OHIF, and we are not going to generate SRs that include OHIF-specific non-standard content.

What functionality is expected to be supported when OHIF encounters an SR that is compliant with the standard but does not have OHIF-specific components? At the moment, point is not rendered, and list of annotations is not shown. It seems that nothing works right now. Is this expected?

**Comment by wayfarer3130** [comment]
Hmm, somehow my reply got lost.
At the moment, OHIF needs the custom annotation format to be able to parse annotations in SR.  That code is only in the annotation parse handling, so writing a fully TID.1500 compliant parsing is on the plan, and would be nice to have submittted.

For now, the only functionality that OHIF has for non-OHIF extension annotations is to display them while in the SR viewport.
This could be extended to allow displaying them in the right hand list panel, but the right way to do that is to just write the parser.

If a parser were written for standards compliant annotations, then any standards compliant annotation would work for all the existing OHIF functionality.  That is, it would appear in the right hand panel, would navigate to the correct image etc.
For creating new annotations/editing them in OHIF, given a parser, it would likely also have a generator so that the full save/load loop exists.

Writing such a parser/generator is not terribly hard - the existing one is ALMOST compliant - there are two changes required, one for recording the annotation type correctly/restoring that value correctly, and the other for handling the annotation text correctly.
There is a small OHIF change to deal with the updated parser as well - that allows recognizing more annotation types.

I would guess a day would allow a single annotation type to be used as a demo/example, followed by a week to go through the remaining annotation types.  There is a require PR on the dcmjs side to support this functionality, already in review.

**Comment by pedrokohler** [comment]
> What does "to display them while in the SR viewport" mean?

I think he means that they'll only be visible, but they won't be hydrated (meaning they won't be editable or interactable at all).

OHIF has a specific viewport for when you "preload" a SR. It can show the measurements but that's before the hydration phase, which will then allow you to see do all of that with the measurement.

When you hydrate the SR (i.e. when you click the load button for the SR), then a more complex process happens behind the scenes and in the end you have a regular StackViewport, and not a SR-specific viewport.

When you just drag the viewport the viewport is a SR-specific one and the measuremrents aren't interactable.

**Comment by igoroctaviano** [comment]
> [@igoroctaviano](https://github.com/igoroctaviano) I put samples of additional planar annotations as follows:
> 
> * gs://tcia-idc-datareviewcoordination/issue-27-prostatex contains a study from ProstateX collection, with an SR that has a single-point SCROORD3D annotation
> * gs://tcia-idc-datareviewcoordination/issue-28-petctlungdx has a study from PET-CT-LungDx collection, with several SCOORD bounding box annotations
> 
> Can you use those in your tests?
> 
> I also imported those into the `projects/idc-tcia/locations/us-central1/datasets/tcia-idc-datareviewcoordination/dicomStores/planar_annotations` DICOM store, but somehow I do not see them in the OHIF sandbox - they files are in the bucket, and there were no import errors, but I don't know if it's an OHIF issue, or a Google Healthcare issue.
> 
> Also here in dropbox: https://www.dropbox.com/s/98rylgt25b2sm9r/planar_annotations.zip?dl=0

@pedrokohler
